src/edu/stanford/nlp/sempre/Formula.java

```diff
@@ -24,6 +24,10 @@
   public abstract LispTree toLispTree();
 
   // Recursively perform some operation on each formula.
+  // Apply to formulas.  If |func| returns false, then recurse on children.
+  public abstract void recurse(Function<Formula, Boolean> func);
```

Comment:
Technically this function is not necessary since you can do everything with map.  But I guess it's nice to have...

I think 'foreach' is a better name, which is more consistent with the other function being called 'map'.